src/extension/vectortiles/src/main/java/org/geoserver/wms/vector/VectorTileMapOutputFormat.java

```diff
@@ -114,6 +114,13 @@ public WebMap produceMap(final WMSMapContent mapContent) throws ServiceException
                                     paintArea,
                                     VectorMapRenderUtils.getMapScale(mapContent, renderingArea),
                                     (FeatureType) featureSource.getSchema()));
+            // buffer is in pixels (style pixels), need to convert to paint area pixels
+            buffer *=
+                    Math.max(
+                            Math.max(
+                                    this.tileBuilderFactory.getOversampleX(),
+                                    this.tileBuilderFactory.getOversampleY()),
```

Comment:
Was getOverSampleX() an getOverSampleY() already part of the API?

There is some support for automatically determining the buffer from the style, or making use of the user suplied value (if the style is "too dynamic").

Reply:
There is actually a layer setting for the buffer, used when the style is too dynamic. This it's part of the layer metadata map, see usage here:
https://github.com/geoserver/geoserver/blob/b76bba7178d3505cd45e318e659c3b32faee4b5e/src/wms/src/main/java/org/geoserver/wms/GetMap.java#L808

Reply:
Also, looking at other code, should this be guarded? E.g., quoting other bits using the oversamplign factors:

```
if (this.tileBuilderFactory.shouldOversampleScale()) {
if (this.tileBuilderFactory.shouldOversampleScale()) {
            paintArea =
                    new Rectangle(
                            this.tileBuilderFactory.getOversampleX() * mapWidth,
                            this.tileBuilderFactory.getOversampleY() * mapHeight);
        }
}
```

I'm also not familiar with this code, see the output format class has a oversampling factor, but the tile builder has two separate ones, not sure how related those are. "git blame" shows they have been coded by two different people, so it may well be they are not properly connected.

Reply:
I've put the guard in, and fixed the byte/pixels typo.

I think the in-the-query-buffer is already being handled;
https://github.com/geoserver/geoserver/blob/master/src/extension/vectortiles/src/main/java/org/geoserver/wms/vector/PipelineBuilder.java#L298

(NOTE: this might want to be multiplied by the oversample factor, but I'm not sure how people are using it).

The oversample factor is hard coded to 16, which is pretty high.  Perhaps we should drop this to 2 or 4?

Reply:
Okay so oversampleX&Y are in part to support the hi-def displays, so a 512x512 "pixel" tile contains 4096x4096 detail (to support those 400 dpi phones do doubt).

So combining these two things: the layer buffer calculated from the style (or supplied by user override) x oversample should be correct.

Only remaining "huh" question is why there are different oversample X and Y...

Reply:
I think the oversample factor was decided to support highres displays? I believe @travislbrundage did it, why not ask him directly? :-D